### ISSUE TITLE
Inflate/Deflate inconsistent across targets - broken on HTML5

### ISSUE THREAD

**Issue by Greg209** [openfl/openfl#892]
Using the following test case it's easy to see the problem in consistency between Flash, CPP, and HTML5 targets which is especially problematic in porting from AS3 to OpenFL where data has created using these methods and requires reloading.

```
package;

import openfl.display.Sprite;
import openfl.utils.ByteArray;

class CompressTest extends Sprite {

    public function new() {
        super();

        var testString = "AABBCCDDEEFFGG";

        var ba = new ByteArray();
        ba.writeUTFBytes( testString );
        printBytes("Clean   ", ba);
        ba.deflate();
        printBytes("Deflated", ba);    
        ba.inflate();
        printBytes("Inflated", ba);
    }

    private function printBytes(hdr:String, b:ByteArray) {
        b.position = 0;

        var byte : Int;
        var str : String = hdr+":";
        while (b.bytesAvailable > 0) {
            byte = b.readUnsignedByte();
            str += StringTools.hex( byte, 2 ) + " ";
        }
        trace( str );
    }
} 
```

This simply creates a ByteArray with static data, deflates it and the inflates it. Between each step, the contents of the ByteArray are printed. The output is as follows (no -Dv2 targets):

Flash:
    Clean   :41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Deflated:73 74 74 72 72 76 76 71 71 75 75 73 73 77 07 00 
    Inflated:41 41 42 42 43 43 44 44 45 45 46 46 47 47 

CPP/Neko:
    Clean   :41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Deflated:78 DA 73 74 74 72 72 76 76 71 71 75 75 73 73 77 07 00 1B 82 03 B9 
    Inflated:41 41 42 42 43 43 44 44 45 45 46 46 47 47 

HTML5 (this one crashes):
    Clean   :41 41 42 42 43 43 44 44 45 45 46 46 47 47  (CompressTest.js, line 4745)
    Deflated:41 41 42 42 43 43 44 44 45 45 46 46 47 47  (CompressTest.js, line 4745)
Error: Invalid data
inflateLoopCompressTest.js:3619
readBytesCompressTest.js:3511
runCompressTest.js:3489
runCompressTest.js:3393
uncompressCompressTest.js:16664
inflateCompressTest.js:16569
CompressTestCompressTest.js:1403
DocumentClassCompressTest.js:1424
createInstanceCompressTest.js:1992
startCompressTest.js:82
initCompressTest.js:62
dispatchCompressTest.js:5166
startCompressTest.js:8239
display_onCompleteCompressTest.js:28349
fCompressTest.js:36834:247
__dispatchEventCompressTest.js:204
__dispatchEventCompressTest.js:303
dispatchEventCompressTest.js:137
onLoadedCompressTest.js:1837
startCompressTest.js:28338
image_onLoadCompressTest.js:8246
fCompressTest.js:36834:247

**Comment by jgranick** [comment]
Could you see if including `<haxelib name="format" />` makes a difference on HTML5?

**Comment by Greg209** [comment]
That is already being used. An exception is thrown otherwise, for html5. I'll check compress/uncompress later to make sure they work but wouldn't solve my issues.

**Comment by Greg209** [comment]
Updated results - including compress/uncompress:

Test string : "AABBCCDDEEFFGG"
Flash:
    Clean   :41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Deflated:73 74 74 72 72 76 76 71 71 75 75 73 73 77 07 00 
    Inflated:41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Clean.......::41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Compressed..::78 DA 73 74 74 72 72 76 76 71 71 75 75 73 73 77 07 00 1B 82 03 B9 
    Uncompressed::41 41 42 42 43 43 44 44 45 45 46 46 47 47 
CPP/Neko:
    Clean   :41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Deflated:78 DA 73 74 74 72 72 76 76 71 71 75 75 73 73 77 07 00 1B 82 03 B9 
    Inflated:41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Clean.......::41 41 42 42 43 43 44 44 45 45 46 46 47 47 
    Compressed..::78 DA 73 74 74 72 72 76 76 71 71 75 75 73 73 77 07 00 1B 82 03 B9 
    Uncompressed::41 41 42 42 43 43 44 44 45 45 46 46 47 47 
HTML5:
    Clean   :41 41 42 42 43 43 44 44 45 45 46 46 47 47  (CompressTest.js, line 4751)
    Deflated:41 41 42 42 43 43 44 44 45 45 46 46 47 47  (CompressTest.js, line 4751)
    Inflated - errors on html5 (CompressTest.js, line 4751)
    Clean.......::41 41 42 42 43 43 44 44 45 45 46 46 47 47  (CompressTest.js, line 4751)
    Compressed..::41 41 42 42 43 43 44 44 45 45 46 46 47 47  (CompressTest.js, line 4751)
    Uncompressed::errors on html5 (CompressTest.js, line 4751)

So basically it seems like html5 deflate/compress doesn't seem to do much, oddly, compress/decompress on Flash matches that of compress/decompress on CPP but also incorrectly inflate & deflate.

**Comment by azrafe7** [comment]
Mmmhh... please anyone correct me if I'm wrong, but from what I could gather it seems `deflate()` is not YET implemented for js. If you look carefully at https://github.com/openfl/lime/blob/80729b21b9725e86b7defeeecd5431954d7442db/lime/utils/ByteArray.hx#L121 and pay attention to the `ifdefs`, there's no path for js (and that's why you get back the original `byteArray` as a result in your test).

I remember there was an issue open on `format` repo to address this (...found it: https://github.com/HaxeFoundation/format/issues/23).

Other inconsistencies between targets are related (I think), to `windowBits`/`algorithm` and wether or not you need/have a header for the compressed bytes (so GZIP vs ZLIB). (if you diff your compressed output between flash and cpp/neko f.e. you'll notice that only some heading and trailing bytes make for the difference, the rest of the bytes are equal).

I've _~recently_ 1:1 ported the wonderful [pako.js](https://github.com/nodeca/pako) to pure haxe (along with _almost_ the entirety of its test suite, but avoiding some things I found confusing, esp. the `to:"string"` feature - if you skim through their closed issues you'll find many users having problems with that, although it's documented). The pako project aims (and seems to be as of now!) to be fully compatible (same input <=> same output) of [zlib](http://www.zlib.net/). I'm sure my effort needs a review (please lend you're help here!), more testing on real world projects, and a round or two of optimizations/refactoring.

... if anyone's interested it's [here](https://github.com/azrafe7/hxPako).

For @Greg209's specific, case using this _should_ work consistently:

```haxe
	var string = "AABBCCDDEEFFGG";
	var u8data = UInt8Array.fromBytes(Bytes.ofString(string));
	var deflated = Pako.deflateRaw(u8data);
	var deflatedStr = [for (i in 0...deflated.length) StringTools.hex(deflated[i], 2)].join(" ");
	trace(deflatedStr);
```

**Comment by Greg209** [comment]
For the time being, I managed to create a solution as in my specific circumstance I'm currently really only interested in the HTML5. For the JS target, the Format lib is used which has it's own InflateImpl implementation however the problem is that when you call ba.inflate in OpenFl, it executes the lime ByteArray uncompress method, which in turn does a :

    var buf = Inflate.run (bytes).getData ();

Where Inflate.run is from the Format lib. This is fine, however that .run method executes :

    return InflateImpl.run(new haxe.io.BytesInput(bytes));

Using the InflateImpl class. The static .run method in InflateImpl creates and instance of InflateImpl as such :

    var inflate = new InflateImpl(i);

This is the problem!!! For an inflate, it should be instantiated as follows :

    var inflate = new InflateImpl( i, false, false );

The two additional false parameters are to indicate whether a 'header' is provided and a 'crc' check. To match the AS3 inflate - these need to be false for the method to inflate correctly.

So basically Format potentially needs modifying to allow parameters to be passed into the .run methods so they can be passed onto the lower calls through to the InfalteImpl creation.

I created a Inflater class which has a static inflate method which in turn directly calls into the InflateImpl, in same way as normal but passes the additional parameters. Here it is if you need it.

	package haxe;

	import format.tools.InflateImpl;
	import flash.utils.CompressionAlgorithm;
	import haxe.io.Bytes;
	import haxe.io.Input;
	import flash.utils.ByteArray;

	class Inflater {
		
		public static function inflate( ba : ByteArray ) {

			var bytes = Bytes.ofData (cast ba.byteView);
			var bytesInput = new haxe.io.BytesInput( bytes );
			var buf = Inflater.inflateRun ( bytesInput );
			var newba = ByteArray.fromBytes( buf );

			return newba;
		}

		public static function inflateRun( i : haxe.io.Input, ?bufsize = 65536 ) {
			var buf = haxe.io.Bytes.alloc(bufsize);
			var output = new haxe.io.BytesBuffer();
			var inflate = new InflateImpl( i, false, false );
			while( true ) {
				var len = inflate.readBytes(buf,0,bufsize);
				output.addBytes(buf,0,len);
				if( len < bufsize )
					break;
			}
			return output.getBytes();
		}
	}

As indicated, this is only for JS/HTML5 targets - the other inconsistencies are still there.

**Comment by lovedna** [comment]
i had the same issue ,uncompress has error in html5 target
i use you code replace uncompress still error in html5 target

**Comment by jgranick** [comment]
This was fixed some time ago :wink:
